dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/BeanTableSchemaSubstitution.java

```diff
@@ -0,0 +1,53 @@
+package io.quarkus.amazon.dynamodb.enhanced.runtime;
+
+import java.beans.PropertyDescriptor;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+
+import software.amazon.awssdk.enhanced.dynamodb.internal.mapper.ObjectConstructor;
+import software.amazon.awssdk.enhanced.dynamodb.mapper.BeanTableSchema;
+
+/**
+ * Although method handles are now supported by GraalVM
+ * (https://github.com/oracle/graal/issues/2761), the {@link ObjectConstructor} (and others)
+ * dynamically creates an efficient suppliers (lambda) for the passed methods.
+ *
+ * <p>
+ * In static, normal java code, this would look like HashMap::new. Since this has to be done with
+ * any (unknown) classes and creating via reflection is slow, this is a workaround to create this
+ * "static" code dynamically at runtime for maximum performance.
+ *
+ * <p>
+ * The problem now is, that this "lambda" is effectively a new class. Since in the native image
+ * the compilation is already done, it is not possible to add additional code at runtime which this
+ * new class is.
+ *
+ * <p>
+ * The solution is, not to create lambdas at runtime but rather to stick to ordinary method
+ * handles. With them we can efficiently create new objects of classes but we still need the wrapper
+ * (supplier, function ,...) for them. The workaround for that is to use the good old inner classes.
+ */
```

Comment:
Is it actually an issue or some theoretical performance problem? Because it looks a bit too involved for me if it's not strictly needed.

Reply:
It is needed but unfortunately, I can't explain it because I don't understand everything myself. I'll give it a try...

The BeanTableSchema class uses java.lang.reflect.* to scan classes annotated with DynamoDbBean for their constructor & methods. The BeanTableSchema.java class is part of the AWS SDK V2. Since by the GraalVM this part was already converted into "native code" reflection doesn't work here anymore.

To make this possible anyway, what BeanTableSchema does is replaced in this repo (BeanTableSchemaSubstitution.java) and redirected to the BeanTableSchemaSubstitutionImplementation by the MethodCallRedirectionVisitor class in the EnhancedProcessor.java.

In EnhancedProcessor.java, the static class MethodCallRedirectionVisitor overrides visitMethod() and redirects method calls to BeanTableSchemaSubstitutionImplementation. This is done with the help of inner classes instead of lambda expressions.

Reply:
Re-reading my explanation many months after I wrote it makes it clear to me that it is not exactly great to understand. And it seems everyone has a general grasp but there is still a some of confusion around what is happening here. I mean I admit that I definitely had my WTF-moments reading my shit.

> Since by the GraalVM this part was already converted into "native code" reflection doesn't work here anymore.

Reflection can be a problem (because everything unneeded is discarded) but we work around that by telling native-image to include it via the `ReflectiveClassBuildItem`s (or rather we tell Quarkus to tell native-image). The real problem are the dynamically generated lambdas which count as classes and they cannot be loaded any more.

I mean you could solve this in a couple of different ways. Like to somehow pre-generate the lambdas and include them in the native-image. Or drop the lambda-magic and use annotation-processors to generate the needed binding code. But for me the path of least resistance was just to substitute the lambda generation with something supported.

> To make this possible anyway, what BeanTableSchema does is replaced in this repo (BeanTableSchemaSubstitution.java) and redirected to the BeanTableSchemaSubstitutionImplementation by the MethodCallRedirectionVisitor class in the EnhancedProcessor.java.

Not sure If I am misunderstanding but: the `MethodCallRedirectionVisitor` is not used for native-image, only Quarkus tests. The annotation `@Substitute` directly instructs native-image what to do. This is not possible with the JVM, so the `MethodCallRedirectionVisitor` is essentially the equivalent on the JVM side. A lot more verbose though. 

As I am thinking about it right now... (If anyone is still confused whats going on here, skip this one paragraph.) 
So I started with the `@Substitute` because I only knew about the native-image problems. When trying to write tests I fell over the test-problems. So I added the Visitior-stuff for the tests. BUT you should just be able to remove the Substitutions and use something like `@BuildStep(onlyIf = NativeBuild.class)` with the Visitors. The reason behind that is that the fix is identical for both native-image and the tests. We already rewrite the bytecode to fix the tests so if we pass that on to native-image, we don't need the `@Substitute` because we already substituted manually.


> Is it actually an issue or some theoretical performance problem? Because it looks a bit too involved for me if it's not strictly needed.

This is the core fix for the `BeanTableSchema`. It would crash at runtime with something along the lines of "Cannot define class at runtime." (referring to the lambda as class). 

---

So actually we are trying to fix *two separate* issues here. The first being the native-image and the second one the tests with Quarkus. Additionally, the `@DynamoDbBean`s are reflectively introspected at runtime, so we automatically "`@RegisterForReflection`" them for the user as convenience; not really as "fix".

- We are substituting the usage of runtime generated lambdas with method handles because they are not supported by native-image.
- We are substituting the usage of runtime generated lambdas because they are wrongly generated by the AWS SDK when used across different classloaders (AWS SDK and user-defined bean classes in separate classloaders). [Nearly correct explanation by @hamburml](https://github.com/quarkiverse/quarkus-amazon-services/pull/255/files/0fe13b465dff6afc6508dc4cfc2da4bf42573f2c#r970879162). Also some [more info here](https://github.com/quarkiverse/quarkus-amazon-services/pull/255#issuecomment-1233347772) and [here](https://github.com/aws/aws-sdk-java-v2/issues/2604).

If needed I could re-formulate a standalone explanation of everything that does not refer to comments. I would need to know how detailed and if in markdown or javadoc.

---

I hope that answers at least some questions? Sorry, I skipped it a little bit. I don't have an overview of who is currently confused where.

Reply:
The longer this takes, the more confused I get....

> Not sure If I am misunderstanding but: the MethodCallRedirectionVisitor is not used for native-image, only Quarkus tests.

My bad, I thought it is connected.

I am honest - I simply do not know what I should do now. Should I try to remove `@Substitute`? In fact I did, I removed the whole `BeanTabelSchema.java` file and currently only the `BeanTableSchemaSubstitutionImplementation` with the methodHandlers exist. I can build it but in my example application I get the following...

```
docker-dynamodb-enhanced-test-1  | 2022-09-20 18:35:58,210 DEBUG [io.ver.ext.web.RoutingContext]
(executor-thread-0) RoutingContext failure (500): org.jboss.resteasy.spi.UnhandledException: 
java.lang.IllegalArgumentException: Failed to generate method handle.
....
docker-dynamodb-enhanced-test-1  | Caused by: java.lang.InternalError: com.oracle.svm.core.jdk.UnsupportedFeatureError: Defining hidden classes at runtime is not supported.
docker-dynamodb-enhanced-test-1  |      at java.lang.invoke.InnerClassLambdaMetafactory.generateInnerClass(InnerClassLambdaMetafactory.java:413)
docker-dynamodb-enhanced-test-1  |      at java.lang.invoke.InnerClassLambdaMetafactory.spinInnerClass(InnerClassLambdaMetafactory.java:315)
docker-dynamodb-enhanced-test-1  |      at java.lang.invoke.InnerClassLambdaMetafactory.buildCallSite(InnerClassLambdaMetafactory.java:228)
docker-dynamodb-enhanced-test-1  |      at java.lang.invoke.LambdaMetafactory.metafactory(LambdaMetafactory.java:341)
docker-dynamodb-enhanced-test-1  |      at software.amazon.awssdk.enhanced.dynamodb.internal.mapper.LambdaToMethodBridgeBuilder.build(LambdaToMethodBridgeBuilder.java:83)
docker-dynamodb-enhanced-test-1  |      ... 37 more
docker-dynamodb-enhanced-test-1  | Caused by: com.oracle.svm.core.jdk.UnsupportedFeatureError: Defining hidden classes at runtime is not supported.
docker-dynamodb-enhanced-test-1  |      at com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:89)
docker-dynamodb-enhanced-test-1  |      at java.lang.ClassLoader.defineClass0(ClassLoader.java:336)
docker-dynamodb-enhanced-test-1  |      at java.lang.System$2.defineClass(System.java:2307)
docker-dynamodb-enhanced-test-1  |      at java.lang.invoke.MethodHandles$Lookup$ClassDefiner.defineClass(MethodHandles.java:2439)
docker-dynamodb-enhanced-test-1  |      at java.lang.invoke.MethodHandles$Lookup$ClassDefiner.defineClassAsLookup(MethodHandles.java:2420)
docker-dynamodb-enhanced-test-1  |      at java.lang.invoke.MethodHandles$Lookup.defineHiddenClass(MethodHandles.java:2127)
docker-dynamodb-enhanced-test-1  |      at java.lang.invoke.InnerClassLambdaMetafactory.generateInnerClass(InnerClassLambdaMetafactory.java:407)


```
> BUT you should just be able to remove the Substitutions and use something like `@BuildStep(onlyIf = NativeBuild.class)` with the Visitors.

I do not understand this. If I add onlyIf = NativeBuild.class to `maybeApplyClassTransformation` or `applyClassTransformation` the whole fix would not work anymore in jvmMode.

Reply:
got it :D I think I just forget to clean the project.